### DISCUSSION TITLE
Timers on ESP32 are liternal not virtual right?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#16494] [General]
Timers on ESP32 are liternal not virtual right?  Timer(1), Timer(2), etc. 4 possible?  Right?  Then why does T = Timer(-1) not generate an exception or return T = None?  ESP8266 virtual timers are supported so Timer(-1) would not generate an exception.

**Comment by Jibun-no-Kage** [comment]
ESP32 16 bit timers of course 4... versus 2 group as 32 bit... I did not quality that in the above.  But still think -1 as timer index or id on ESP32 should trip an exception or something... it appears it is accepted, but never fires.  Even... T.init() does not freak out on ESP32 if you happen to use a virtual id or index.

**Reply by robert-hh** [reply]
There was a recent change on how timer numbers are handled. Try the most recent v1.25 preview. With 1.24, Timer(-1) should be the same as Timer(3) because only the two lowest bits of the number  were used to choose a timer.

**Reply by robert-hh** [reply]
Log with a recent preview: 
```
MicroPython v1.25.0-preview.150.g7ff4d04bb on 2024-12-21; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> from machine import Timer
>>> t=Timer(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid Timer number
```

**Comment by Jibun-no-Kage** [comment]
What is weird is that I never got that error... I check for T=None it passed T was not None.  But the timer  never fired.  Using 1.24.1 btw.  I changed the timer id to 1,2,3,4 fired just fine.  but as -1 never fired.  Interesting.
